content/posts/gitops-series-part-1.md

```diff
@@ -0,0 +1,616 @@
+
+## Setting up Cert Manager
+
+Now we have our ingress controller which has given us a public ip address and we have setup a DNS record to point to the address. We could go ahead and install Core or CJD at this point if we wanted, but we might as well setup SSL certificates to make things more secure.
+
+We're going to use [Let's Encrypt](https://letsencrypt.org/) Certificate Authority in order to generate the certs. To do this in an easy and automated fashion, we'll use [cert-manager](https://github.com/jetstack/cert-manager). 
+
+Installing it is pretty easy, and you can find the most update instructions [here](https://docs.cert-manager.io/en/latest/getting-started/install/kubernetes.html).
+
+```bash
+kubectl create namespace cert-manager
+kubectl label namespace cert-manager certmanager.k8s.io/disable-validation=true
+kubectl apply -f https://github.com/jetstack/cert-manager/releases/download/v0.8.0/cert-manager.yaml
+```
+
+The above commands create a namespace for cert-manager and then deploy the cert-manager application in there. 
+
+Next we'll need to create some cert issuers to allow us to actually generate the certs. We'll create two of them in the cert-manager directory.
+
+`cert-manager/staging-issuer.yaml`:
+```yaml
+   apiVersion: certmanager.k8s.io/v1alpha1
+   kind: Issuer
+   metadata:
+     name: letsencrypt-staging
+   spec:
+     acme:
+       # The ACME server URL
+       server: https://acme-staging-v02.api.letsencrypt.org/directory
+       # Email address used for ACME registration
+       email: myemail@example.com # change this
+       # Name of a secret used to store the ACME account private key
+       privateKeySecretRef:
+         name: letsencrypt-staging
+       # Enable the HTTP-01 challenge provider
+       http01: {}
+```
+
+`cert-manager/production-issuer.yaml`:
+```yaml
+   apiVersion: certmanager.k8s.io/v1alpha1
+   kind: Issuer
+   metadata:
+     name: letsencrypt-prod
+   spec:
+     acme:
+       # The ACME server URL
+       server: https://acme-v02.api.letsencrypt.org/directory
+       # Email address used for ACME registration
+       email: myemail@example.com
+       # Name of a secret used to store the ACME account private key
+       privateKeySecretRef:
+         name: letsencrypt-prod
+       # Enable the HTTP-01 challenge provider
+       http01: {}
+```
+
+The reason we have two of these is because Let's Encrypt has a rate-limiter on how often you can generate certificates. So while you are experimenting with things, it is safer to use the staging issuer. When things are all sorted, you can switch to the production-issuer.
+
+Go ahead and apply these two issuers with:
+
+```bash
+kubectl apply -f cert-manager/staging-issuer.yaml
+kubectl apply -f cert-manager/production-issuer.yaml
+```
+
+Now when we create an ingress for our applications, we can add some metadata to the ingress definition and the certificates will automatically be generated and stored in secrets.
```

Comment:
"stored in secrets." or "stored as [K8s Secrets](https://kubernetes.io/docs/concepts/configuration/secret/)"